File: src/module/api/download.py
from fastapi import Depends, HTTPException, status

from module.manager import SeasonCollector
from module.models import BangumiData
from module.models.api import RssLink
from module.rss import analyser
from module.security import get_current_user

from .config import router


@router.post("/api/v1/download/analysis", tags=["download"])
async def analysis(link: RssLink, current_user=Depends(get_current_user)):
    if not current_user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="invalid token"
        )
    data = analyser.link_to_data(link.rss_link)
    if data:
        return data
    else:
        return {"status": "Failed to parse link"}


@router.post("/api/v1/download/collection", tags=["download"])
async def download_collection(
    data: BangumiData, current_user=Depends(get_current_user)
):
    if not current_user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="invalid token"
        )
    if data:
        with SeasonCollector() as collector:
            if collector.collect_season(data, data.rss_link[0], proxy=True):
                return {"status": "Success"}
            else:
                return {"status": "Failed to add torrent"}
    else:
        return {"status": "Failed to parse link"}


@router.post("/api/v1/download/subscribe", tags=["download"])
async def subscribe(data: BangumiData, current_user=Depends(get_current_user)):
    if not current_user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="invalid token"
        )
    if data:
        with SeasonCollector() as collector:
            collector.subscribe_season(data)
        return {"status": "Success"}
    else:
        return {"status": "Failed to parse link"}

File: src/api/download.ts
import type { BangumiRule } from '#/bangumi';

interface Status {
  status: 'Success';
}

interface AnalysisError {
  status: 'Failed to parse link';
}

export const apiDownload = {
  /**
   * 解析 RSS 链接
   * @param rss_link - RSS 链接
   */
  async analysis(rss_link: string) {
    const fetchResult = createEventHook<BangumiRule>();
    const fetchError = createEventHook<AnalysisError>();

    axios
      .post<any>('api/v1/download/analysis', {
        rss_link,
      })
      .then(({ data }) => {
        if (data.status) {
          fetchError.trigger(data as AnalysisError);
        } else {
          fetchResult.trigger(data as BangumiRule);
        }
      });

    return {
      onResult: fetchResult.on,
      onError: fetchError.on,
    };
  },

  /**
   * 旧番
   * @param bangumiData - Bangumi 数据
   */
  async collection(bangumiData: BangumiRule) {
    const { data } = await axios.post<Status>(
      'api/v1/download/collection',
      bangumiData
    );
    return data.status === 'Success';
  },

  /**
   * 新番
   * @param bangumiData - Bangumi 数据
   */
  async subscribe(bangumiData: BangumiRule) {
    const { data } = await axios.post<Status>(
      'api/v1/download/subscribe',
      bangumiData
    );
    return data.status === 'Success';
  },
};
